### DISCUSSION TITLE
has signature for Assembly.cacheResidual/Jacobian() changed recently?

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#24987] [Q&A General] [answered]
I recently updated my local copy of the framework and broke my app. I have a NodalKernel that has calls to `_assembly.cacheResidual/Jacobian()` to add some values to specific rows in the matrix. There were previously `_vector_tags` and `_matrix_tags` member variables that I blindly passed to the functions as the last argument.

After updating my framework and re-compiling my app, I get errors like: `error: 'std::set<unsigned int> TaggingInterface::_vector_tags' is private within this context`. How can I fix?

**Comment by lindsayad** [comment, marked as answer]
It has changed recently. I would need to see your code snippets to know the exact right sub, but here is what things look like now in `NodalKernel`
```c++
void
NodalKernel::computeResidual()
{
  if (_var.isNodalDefined())
  {
    const dof_id_type & dof_idx = _var.nodalDofIndex();
    _qp = 0;
    const Real res = computeQpResidual();
    addResiduals(_assembly,
                 std::array<Real, 1>{{res}},
                 std::array<dof_id_type, 1>{{dof_idx}},
                 _var.scalingFactor());

    if (_has_save_in)
      for (const auto & var : _save_in)
        var->sys().solution().add(var->nodalDofIndex(), res);
  }
}

void
NodalKernel::computeJacobian()
{
  if (_var.isNodalDefined())
  {
    _qp = 0;
    const Real cached_val = computeQpJacobian();
    const dof_id_type cached_row = _var.nodalDofIndex();

    addJacobianElement(_assembly, cached_val, cached_row, cached_row, _var.scalingFactor());

    if (_has_diag_save_in)
    {
      Threads::spin_mutex::scoped_lock lock(Threads::spin_mtx);
      for (const auto & var : _diag_save_in)
        var->sys().solution().add(var->nodalDofIndex(), cached_val);
    }
  }
}

void
NodalKernel::computeOffDiagJacobian(const unsigned int jvar_num)
{
  const auto & jvar = getVariable(jvar_num);

  if (_var.isNodalDefined())
  {
    if (jvar.number() == _var.number())
      computeJacobian();
    else
    {
      _qp = 0;
      const Real cached_val = computeQpOffDiagJacobian(jvar.number());
      const dof_id_type cached_row = _var.nodalDofIndex();

      // Note: this only works for equal order Lagrange variables...
      const dof_id_type cached_col = _current_node->dof_number(_sys.number(), jvar.number(), 0);

      addJacobianElement(_assembly, cached_val, cached_row, cached_col, _var.scalingFactor());
    }
  }
}
```

**Reply by jessecarterMOOSE** [reply]
@lindsayad ok i lied, I'm using a `NodalScalarKernel`. Not sure if that matters. 

Where are `addResiduals` and `addJacobianElement` defined? I assume these have the same effect as what `cacheResidual/Jacobian` did previously?

The Jacobian calls look similar in that you still pass the row, column, and value. The scaling factor is not applied here but is instead passed as well as the assembly reference. Seens straightforward.

How about the Residual calls? Previously I had `_assembly.cacheResidual(_var.dofIndices()[0], computeQpResidual()*_var.scalingFactor(), _vector_tags)`. What's this `std::array` business?

**Reply by lindsayad** [reply]
It allows us to use the same API for both a single residual and a dynamically sized vector of residuals

**Reply by lindsayad** [reply]
> Where are addResiduals and addJacobianElement defined? 

You can see all the available APIs in `TaggingInterface`

**Reply by jessecarterMOOSE** [reply]
What's the `_qp = 0` doing in there? Doesn't seem to be used.

**Reply by lindsayad** [reply]
It's assumed that's is used in the derived class

**Reply by jessecarterMOOSE** [reply]
OK, I think I got this figured out. Thanks for your help @lindsayad . I think the change is for the better since I don't have to bother with tags in my kernel and don't need to remember to apply the scaling factor.
